install-mimic.1
.\" SPDX-FileCopyrightText: Peter Pentchev <roam@ringlet.net>
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.Dd June 2, 2015
.Dt INSTALL-MIMIC 1
.Os
.Sh NAME
.Nm install-mimic
.Nd overwrite files preserving the destination's ownership and access mode
.Sh SYNOPSIS
.Nm
.Op Fl v
.Op Fl r Ar reffile
.Ar srcfile
.Ar dstfile
.Nm
.Op Fl v
.Op Fl r Ar reffile
.Ar file1
.Op Ar file2...
.Ar directory
.Nm
.Fl V | Fl h
.Sh DESCRIPTION
The
.Nm
utility copies the specified files to the specified destination (file or
directory) similarly to
.Xr install 1 ,
but it preserves the ownership and access mode of the destination files.
This is useful when updating files that have already been installed with
locally modified copies that may be owned by the current user and not by
the desired owner of the destination file (e.g. root).
.Pp
Options:
.Bl -tag -width indent
.It Fl h
Display program usage information and exit.
.It Fl r
Specify a file to obtain the ownership and access information from;
useful when installing a new file that should be similar to other files in
the same location.
.It Fl V
Display program version information and exit.
.It Fl v
Verbose operation; display diagnostic output.
.El
.Sh RETURN VALUES
.Ex -std
.Sh ENVIRONMENT
The
.Nm
utility's operation is currently not directly affected by its environment.
.Sh FILES
The
.Nm
utility's operation is currently not directly affected by any files.
.Sh EXAMPLES
Overwrite a system file with a local copy:
.Pp
.Dl install-mimic ./install-mimic.pl /usr/bin/install-mimic
.Pp
Overwrite several files with local copies with the same name:
.Pp
.Dl install-mimic cinder/*.py /usr/lib/python2.7/dist-packages/cinder/
.Pp
Install a new file similar to a system file:
.Pp
.Dl install-mimic -v -r /usr/bin/install-mimic install-none /usr/bin/
.Pp
.Sh HISTORY
The
.Nm
utility was written by
.An Peter Pentchev
in 2015.
.Sh AUTHORS
.An Peter Penchev
.Aq roam@ringlet.net


docs/changes.md
<!--
SPDX-FileCopyrightText: Peter Pentchev <roam@ringlet.net>
SPDX-License-Identifier: BSD-2-Clause
-->

# Changelog

All notable changes to the install-mimic project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.4.1] - 2024-02-26

### Additions

- Add an EditorConfig definitions file.
- Start some MkDocs-based documentation.
- Add a Tox configuration file for running the `reuse` SPDX check tool and
  building the documentation.
- Rust implementation:
    - add the `run-clippy.sh` tool for running diagnostic checks

### Other changes

- Switch to yearless copyright notices with my e-mail address.
- Use SPDX copyright and license tags.
- Rust implementation:
    - rework it, bringing it up to date with Rust edition 2021 and with
      other changes since it was introduced
    - use the `shell-words` crate
    - use the `clap` crate for command-line parsing
    - use `anyhow` instead of `expect-exit`
    - make the path to `cargo` configurable in the Makefile

## [0.4.0] - 2018-05-04

### Additions

- Add the `--help` and `--version` long options.
- Add the `--features` long option.

## [0.3.1] - 2017-09-29

### Fixes

- In testing, get the file group from a new file created in
  the test directory to fix the case of enforced setgid directories.
- Create the test temporary directory in the system's temporary path
  to avoid future weird situations like the setgid case.

## [0.3.0] - 2017-02-27

### Fixes

- Fix a memory allocation bug in the C implementation leading to
  destination filename corruption when the target specified on
  the command line is a directory.

### Additions

- Add a Rust implementation.

## [0.2.0] - 2016-06-29

### Fixes

- Explicitly test the Perl 5 implementation in the "test" target.
- Let the tests continue if an expected file was not created.

### Additions

- Add tests for the -r reffile and -v command-line options.
- Add a C implementation.

## [0.1.1] - 2016-06-28

### Additions

- Add the internal "dist" target for creating distribution tarballs.
- Add a test suite.
- Add a Travis CI configuration file and a cpanfile.

### Other changes

- Reorder the functions a bit to avoid prototype declarations.
- Make the usage() function fatal by default.
- Move development from GitLab to GitHub.
- Switch the homepage URL to HTTPS.

## [0.1.0] - 2015-06-02

### Started

- First public release.

[Unreleased]: https://github.com/ppentchev/install-mimic/compare/release%2F0.4.1...master
[0.4.1]: https://github.com/ppentchev/install-mimic/compare/release%2F0.4.0...release%2F0.4.1
[0.4.0]: https://github.com/ppentchev/install-mimic/compare/release%2F0.3.1...release%2F0.4.0
[0.3.1]: https://github.com/ppentchev/install-mimic/compare/release%2F0.3.0...release%2F0.3.1
[0.3.0]: https://github.com/ppentchev/install-mimic/compare/release%2F0.2.0...release%2F0.3.0
[0.2.0]: https://github.com/ppentchev/install-mimic/compare/release%2F0.1.1...release%2F0.2.0
[0.1.1]: https://github.com/ppentchev/install-mimic/compare/release%2F0.1.0...release%2F0.1.1
[0.1.0]: https://github.com/ppentchev/install-mimic/releases/tag/release%2F0.1.0


install-mimic.c
/*-
 * SPDX-FileCopyrightText: Peter Pentchev <roam@ringlet.net>
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>

#include <err.h>
#include <errno.h>
#include <inttypes.h>
#include <libgen.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifndef __printflike
#if defined(__GNUC__) && __GNUC__ >= 3
#define __printflike(x, y)	__attribute__((format(printf, (x), (y))))
#else
#define __printflike(x, y)
#endif
#endif

#define VERSION_STRING	"0.4.1"

static bool		verbose;

static void
version(void)
{
	puts("install-mimic " VERSION_STRING);
}

static void
usage(const bool _ferr)
{
	const char * const s =
	    "Usage:\tinstall-mimic [-v] [-r reffile] srcfile dstfile\n"
	    "\tinstall-mimic [-v] [-r reffile] file1 [file2...] directory\n"
	    "\tinstall-mimic -V | --version | -h | --help\n"
	    "\tinstall-mimic --features\n"
	    "\n"
	    "\t-h\tdisplay program usage information and exit\n"
	    "\t-r\tspecify a reference file to obtain the information from\n"
	    "\t-V\tdisplay program version information and exit\n"
	    "\t-v\tverbose operation; display diagnostic output\n";

	fprintf(_ferr? stderr: stdout, "%s", s);
	if (_ferr)
		exit(1);
}

static void __printflike(3, 4)
snprintf_check(char * const buf, const size_t sz, const char * const fmt, ...)
{
	va_list v;

	va_start(v, fmt);
	const int n = vsnprintf(buf, sz, fmt, v);
	va_end(v);
	if (n < 0 || (size_t)n >= sz)
		errx(1, "Internal error, could not fit '%s'-formatted output "
		    "into %zu characters", fmt, sz);
}

static void
debug_cmd(const char * const *cmd)
{
	if (!verbose)
		return;

	bool first = true;
	while (*cmd != NULL) {
		printf("%s%s", first? "": " ", *cmd);
		first = false;
		cmd++;
	}
	putchar('\n');
}

static void
check_wait_result(const pid_t pid, const int stat, const pid_t expected, const char * const progname)
{
	if (pid != expected)
		errx(1, "Waiting for %s: expected pid %ld, got %ld", progname, (long)expected, (long)pid);
	else if (WIFEXITED(stat) && WEXITSTATUS(stat) != 0)
		errx(1, "Child %s (pid %ld) exited with code %d", progname, (long)pid, WEXITSTATUS(stat));
	else if (WIFSIGNALED(stat))
		errx(1, "Child %s (pid %ld) was killed by signal %d", progname, (long)pid, WTERMSIG(stat));
	else if (WIFSTOPPED(stat))
		errx(1, "Child %s (pid %ld) was stopped by signal %d", progname, (long)pid, WSTOPSIG(stat));
	else if (!WIFEXITED(stat))
		errx(1, "Child %s (pid %ld) neither exited nor was killed or stopped; what in the world does wait(2) status %d mean?!", progname, (long)pid, stat);
}

static void
install_mimic(const char * const src, const char * const dst,
		const char * const ref)
{
	const char * const refname = ref != NULL? ref: dst;
	struct stat sb;
	if (stat(refname, &sb) == -1)
		err(1, "Could not stat %s", refname);

	char owner[20], group[20], mode[6];
	snprintf_check(owner, sizeof(owner), "%jd", (intmax_t)sb.st_uid);
	snprintf_check(group, sizeof(group), "%jd", (intmax_t)sb.st_gid);
	snprintf_check(mode, sizeof(mode), "%04o", sb.st_mode & 07777);

	const char * const cmd[] = {
		"install", "-c", "-o", owner, "-g", group, "-m", mode,
		"--", src, dst, NULL
	};
	debug_cmd(cmd);

	const pid_t pid = fork();
	if (pid == -1) {
		err(1, "Could not fork for install(1)");
	} else if (pid == 0) {
		execvp("install", (char * const *)cmd);
		err(1, "Could not execute install(1)");
	}
	int status;
	const pid_t wpid = waitpid(pid, &status, 0);
	if (wpid == -1)
		err(1, "Could not wait for the install(1) child");
	check_wait_result(wpid, status, pid, "install(1)");
}

static bool is_dir(const char * const path, const bool has_ref)
{
	struct stat st;
	if (stat(path, &st) == -1) {
		if (errno != ENOENT || !has_ref)
			err(1, "Could not check whether %s is a directory", path);
		return (false);
	}
	return (S_ISDIR(st.st_mode));
}

int
main(int argc, char * const argv[])
{
	bool features = false, hflag = false, Vflag = false;
	const char *ref = NULL;
	int ch;
	while (ch = getopt(argc, argv, "hr:Vv-:"), ch != -1)
		switch (ch) {
			case 'h':
				hflag = true;
				break;

			case 'r':
				ref = optarg;
				break;

			case 'V':
				Vflag = true;
				break;

			case 'v':
				verbose = true;
				break;

			case '-':
				if (strcmp(optarg, "features") == 0)
					features = true;
				else if (strcmp(optarg, "help") == 0)
					hflag = true;
				else if (strcmp(optarg, "version") == 0)
					Vflag = true;
				else
					usage(1);
				break;

			default:
				usage(1);
				/* NOTREACHED */
		}
	if (Vflag)
		version();
	if (hflag)
		usage(false);
	if (features)
		puts("Features: install-mimic=" VERSION_STRING);
	if (Vflag || hflag || features)
		return (0);

	argc -= optind;
	argv += optind;
	if (argc < 2)
		usage(true);

	const char * const last_dir = argv[argc - 1];
	if (is_dir(last_dir, ref != NULL)) {
		const size_t last_dir_len = strlen(last_dir);
		const bool has_slash = last_dir_len == 0? false:
		    last_dir[last_dir_len - 1] == '/';
		char *basebuf = NULL, *fullbuf = NULL;

		for (int i = 0; i < argc - 1; i++)
		{
			/* Sigh, basename(3) may modify its argument */
			const char * const src = argv[i];
			const size_t sz = strlen(src) + 1;
			basebuf = realloc(basebuf, sz);
			if (basebuf == NULL)
				err(1, "Could not allocate %zu bytes", sz);
			snprintf_check(basebuf, sz, "%s", src);
			const char * const base = basename(basebuf);

			const size_t fullsz = last_dir_len + !has_slash +
			    strlen(base) + 1;
			fullbuf = realloc(fullbuf, fullsz);
			if (fullbuf == NULL)
				err(1, "Could not allocate %zu bytes", fullsz);
			snprintf_check(fullbuf, fullsz, "%s%s%s",
			    last_dir, has_slash? "": "/", base);

			install_mimic(src, fullbuf, ref);
		}
		free(basebuf);
		free(fullbuf);
	} else {
		if (argc != 2)
			usage(true);
		install_mimic(argv[0], argv[1], ref);
	}
	return (0);
}
